src/shared/module-util.c

```diff
@@ -50,9 +51,25 @@ int module_load_and_warn(struct kmod_ctx *ctx, const char *module, bool verbose)
                                          "Inserted module '%s'", kmod_module_get_name(mod));
                         else if (err == KMOD_PROBE_APPLY_BLACKLIST)
                                 log_full(verbose ? LOG_INFO : LOG_DEBUG,
-                                         "Module '%s' is deny-listed", kmod_module_get_name(mod));
-                        else {
-                                assert(err < 0);
+                                         "Module '%s' is deny-listed (by kmod)", kmod_module_get_name(mod));
+                        else if (err == -EPERM) {
+                                STRV_FOREACH(i, blacklist) {
+                                        if (streq(*i, kmod_module_get_name(mod))) {
+                                                log_full(verbose ? LOG_INFO : LOG_DEBUG,
+                                                         "Module '%s' is deny-listed (by kernel)",
+                                                         kmod_module_get_name(mod));
+                                                r = -err; /* shouldn't be considered as a failure, so we negate the return value */
+                                        }
+                                }
+                                if (r <= 0) {
+                                        log_full_errno(!verbose ? LOG_DEBUG : LOG_ERR,
+                                                       err,
+                                                       "Failed to insert module '%s': %m",
+                                                       kmod_module_get_name(mod));
+                                        r = err;
+                                }
```

Comment:
This is mostly a duplicate of below. How about the following?
```c
int module_load_and_warn(struct kmod_ctx *ctx, const char *module, bool verbose) {
        _cleanup_strv_free_ char **blacklist = NULL;
        bool blacklist_parsed = false;
        ...
        kmod_list_foreach(itr, modlist) {
                ...
                default:
                        err = kmod_module_probe_insert_module(mod, probe_flags,
                                                              NULL, NULL, NULL, NULL);
                        if (err == 0)
                                log_full(verbose ? LOG_INFO : LOG_DEBUG,
                                         "Inserted module '%s'", kmod_module_get_name(mod));
                        else if (err == KMOD_PROBE_APPLY_BLACKLIST)
                                log_full(verbose ? LOG_INFO : LOG_DEBUG,
                                         "Module '%s' is deny-listed (by kmod)", kmod_module_get_name(mod));
                        else {
                                assert(err < 0);
                                if (err == -EPERM) {
                                        if (!blacklist_parsed) {
                                                r = proc_cmdline_parse(...);
                                                if (r < 0)
                                                        log_full_errno(!verbose ? LOG_DEBUG : LOG_ERR, r, "Failed to parse kernel command line, ignoring: %m");
                                                blacklist_parsed = true;
                                        }
                                        if (strv_contains(blacklist, kmod_module_get_name(mod))) {
                                                log_full_errno(...);
                                                continue;
                                        }
                                }
                                log_full_errno(...);
                                if (!IN_SET(err, -ENODEV, -ENOENT))
                                        r = err;
                        }
```

Reply:
Mostly done expect that I don't know how to use a non-static variable in `parse_proc_cmdline_item` and its callees. If using a static variable is fine, then we're OK.